Create map_iter_next() from core of dict_iter_next(), make use where appropriate

tomlogic opened the pull request:
I noticed a common code pattern for walking the elements of a map, saw a helper function specific to `mp_obj_dict_t`, and decided to generalize it for all maps.  See issue #3105.

I have only made changes to core files, but this should work for `extmod/moductypes.c`, `extmod/moduselect.c`, and `esp8266/modnetwork.c` as well, and I can update them for this pull request if desired.

tomlogic committed:
py/map: Factor new mp_map_iter_next() from dict_iter_next().

issue #3105

tomlogic committed:
py/map: Make use of new mp_map_iter_next().

Replaces looping with an index and checking MP_MAP_SLOT_IS_FILLED().

dpgeorge commented:
I checked the change in code size for this PR and it's as follows:
```
   bare-arm:   -20
    minimal:   -16
   unix x64:   -24
unix nanbox:  -336
     stmhal:   -64
    esp8266:  -208
     cc3200:   -80
```
Not as much as I would have expected, I guess it's because the MP_MAP_SLOT_IS_FILLED macro is actually quite efficient to evaluate.

There is a regression introduced with the dur() builtin.

On the one hand this patch is good because it 1) reduces code size; 2) factorises a common helper to a single function which makes the code easier to read/maintain.  On the other hand this is a deoptimisation and will make the code less efficient, and more importantly it may in some cases increase stack usage because there is the address of a variable being taken.  The stack usage of all the affected functions needs to be looked at in detail to see what the impact is.

tomlogic commented:
A bit confused about the dir() builtint failure and wondering if it's an odd compiler optimization.  When I build the Windows version with MinGW and gcc 4.8.1, it passes all tests, and manually running the failing test shows it passing.

And the Travis build above reported the following:

```
The command "ls -l minimal/build/firmware.bin" exited with 0.
0.01s$ tools/check_code_size.sh
Old size: 77464 new size: 77388
```

Savings of 76 bytes on minimal.

dpgeorge commented on py/modbuiltins.c:
```diff
@@ -225,20 +225,16 @@ STATIC mp_obj_t mp_builtin_dir(size_t n_args, const mp_obj_t *args) {
+            mp_obj_list_append(dir, elem->key);
         }
     }
-    if (members != NULL) {
```
This check still needs to be here.

tomlogic replied:
I moved that check down into `mp_map_iter_next()` in case there were other instances that could save a `NULL` check.  It must be a compiler optimization issue, because the test passed on my build.  I'll pull it out of the inner function and put it back into `py/modbuiltins.c`.

dpgeorge commented:
> Savings of 76 bytes on minimal.

Travis builds minimal for Cortex M arch, but I reported above the build of minimal for x86; sorry it's a bit confusing.

pfalcon commented on py/objdict.c:
```diff
@@ -201,7 +186,7 @@ typedef struct _mp_obj_dict_it_t {
 STATIC mp_obj_t dict_it_iternext(mp_obj_t self_in) {
     mp_obj_dict_it_t *self = MP_OBJ_TO_PTR(self_in);
-    mp_map_elem_t *next = dict_iter_next(MP_OBJ_TO_PTR(self->dict), &self->cur);
+    mp_map_elem_t *next = dict_iter_next((mp_obj_dict_t *)MP_OBJ_TO_PTR(self->dict), &self->cur);
```
Should not be space inbetween: `mp_obj_dict_t *`.

pfalcon commented:
> On the other hand this is a deoptimisation and will make the code less efficient, and more importantly it may in some cases increase stack usage because there is the address of a variable being taken.

I'd say it's concerning then. Generally, mapping handling is 2nd biggest performance inefficiency of MicroPython after GC, which doesn't scale to big mapping sizes. Code size is important, but deoptimizing mapping handling for it doesn't seem like a good compromise.

tomlogic commented:
I believe I found the problem and fixed it.  In the previous version, `mp_map_iter_next()` dereferenced the map object in the `max = map->alloc` assignment before checking `map` against NULL.  Granted we still need to discuss the performance/stack impact of this change, probably after your 1.9 work.

tomlogic committed:
py/map: Calculate element's address once in mp_map_iter_next().

In my builds, IAR's ARM compiler ignored the "inline" attribute of MP_MAP_SLOT_IS_FILLED() and built it as a separate function.  This commit cleans up the declaration of "max" and calculates the element address once for checking whether the slot is filled and as the return value.

dpgeorge commented:
@tomlogic any update on investigating performance/stack usage here?  I guess the first step would be to rebase this PR on the top of current master.

But, otherwise, since the code savings were not very much this PR could be filed under "not worth it".

tomlogic commented:
@dpgeorge Actually, I decided to abandon this patch on a recent merge with upstream and did not run any performance tests.  Based on pushback I got to other PRs from the same timeframe, it didn't feel it was worth advocating for the change.

I did like that we saved code (granted, not much) and simplified some of the calling code (good from a maintenance standpoint).  If I were to revisit it, what sort of information would you want to see?  A test case where we create a large map or dict and then time a large number of lookups on it?

dpgeorge commented:
> If I were to revisit it, what sort of information would you want to see?

I have a local tool for checking the stack usage of the code (ARM Thumb only).  I ran it on this PR (merged at v1.9) for the stmhal port with arm-none-eabi-gcc 7.3.0.  The result is the following increase in stack usage for this PR:
```
mp_help_add_from_map  16 -> 24 bytes (+8)
mp_import_all         24 -> 32 bytes (+8)
mp_obj_instance_attr  64 -> 72 bytes (+8)
```
That's not as bad an increase as I would have guessed.

So I'd say it's still interesting to consider this patch, but the deciding factor would be any changes in performance.

The problem is we don't have a comprehensive way of measuring absolute performance (tests/bench are to test relative speed of various ways of doing the same thing in Python).  Up until now we just relied on pystone, and it gives a decent measure of performance, but not good enough to evaluate a PR like this one.

Until we have a good benchmarking suite I guess we just put this PR on hold.
